File: 1870-minimum-speed-to-arrive-on-time/1870-minimum-speed-to-arrive-on-time.java
class Solution {
    public int minSpeedOnTime(int[] dist, double hour) {
        int ls=1, hs=10000000, ans= (int)1e7+1;
        while(ls<=hs){
            int s= ls+ ((hs-ls)/2);
            double tt=0;
            for(int i: dist){
                tt= Math.ceil(tt);
                tt+= (1.0*i/s);
                if(tt>hour) break;
            }
            if(tt<= hour){
                ans=s;
                hs=s-1;
            }else ls=s+1;
        }
        return (ans>(int)1e7)? -1 : ans;
    }
}

File: 1870-minimum-speed-to-arrive-on-time/1870-minimum-speed-to-arrive-on-time.cpp
class Solution {
public:
    bool isPossible(vector<int> &dist, double hour, int speed){
        double timeTaken=0;
        for(int i=0; i<dist.size()-1; i++){
            timeTaken+= ceil(1.0* dist[i]/speed);
        }
        timeTaken+= (1.0 *dist.back()/speed);
        return (hour>=timeTaken);
    }
    int minSpeedOnTime(vector<int>& dist, double hour) {
        int lowSpeed=1, highSpeed = 10000000, ans=INT_MAX;
        while(lowSpeed<=highSpeed){
            int currSpeed= lowSpeed+((highSpeed-lowSpeed)/2);
            if(isPossible(dist, hour, currSpeed)){
                ans= currSpeed;
                highSpeed= currSpeed-1;
            }else{
                lowSpeed= currSpeed+1;
            }
        }
        return (ans==INT_MAX)? -1: ans;
    }
};